parlai/agents/memnn/memnn.py

```diff
@@ -100,32 +108,38 @@ def __init__(self, opt, shared=None):
             if opt.get('model_file') and os.path.isfile(opt['model_file']):
                 print('Loading existing model parameters from ' + opt['model_file'])
                 self.load(opt['model_file'])
+        else:       
+            self.answers = shared['answers']
 
+        self.history = {}
         self.episode_done = True
         self.last_cands, self.last_cands_list = None, None
         super().__init__(opt, shared)
 
     def share(self):
         shared = super().share()
+        shared['answers'] = self.answers
         return shared
 
     def observe(self, observation):
-        observation = copy.copy(observation)
-        if not self.episode_done:
-            # if the last example wasn't the end of an episode, then we need to
-            # recall what was said in that example
-            prev_dialogue = self.observation['text'] if self.observation is not None else ''
-
-            # append answer given in the previous example to the dialog 
-            if 'eval_labels' in self.observation:
-                prev_dialogue += '\n' + random.choice(self.observation['eval_labels'])
-            elif 'labels' in self.observation:
-                prev_dialogue += '\n' + random.choice(self.observation['labels'])
-
-            observation['text'] = prev_dialogue + '\n' + observation['text']
-        self.observation = observation
+        """Save observation for act.
+        If multiple observations are from the same episode, concatenate them.
+        """
         self.episode_done = observation['episode_done']
-        return observation
+        # shallow copy observation (deep copy can be expensive)
+        obs = observation.copy()
+        batch_idx = self.opt.get('batchindex', 0)
+        
+        obs['text'] = '\n'.join(maintain_dialog_history(
```

Comment:
i think this can actually be much faster if you convert to vectors only once and keep the vector representation.. that's what seq2seq version does

Reply:
I agree, but it will require more changes throughout the file, for example, in the input parse(self, text) since the model uses '\n' to separate query from the context. Can we assume that every context sentence ends with '.' while query with '?'. Maybe leave it for a separate pull request?

Reply:
no we can't assume that, this isn't just for question answering.. i think the \n split is good...